[AutoDiff] Fix forward-mode crashes related to tangent buffers

efremale opened the pull request:
Fixes TF-984, SR-13447.

Fixes foward-mode crashed related to:
- Missing tangent buffers for non-wrt `inout` parameters
- Tangent buffers not being initialized due to the corresponding
original buffer intialization instructions being non-active
- Non-varied indirect results not being initialized
- `emitDestroyValue` crashes due to `TangentVector` value category mismatch

dan-zheng commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -1283,6 +1341,14 @@ class JVPCloner::Implementation final
+      // Assert that local allocations have at least one use.
+      // Buffers should not be allocated needlessly.
+      assert(!alloc->use_empty());
+      diffBuilder.createDeallocStack(diffLoc, alloc);
```
I wonder why `destroy_addr` for local allocations isn't needed here, since the corresponding logic in `PullbackCloner` does generate `destroy_addr` for local allocations.

https://github.com/apple/swift/blob/a5ca83a8b2d499ff1dd0bd7693e779a9ce3aca8c/lib/SILOptimizer/Differentiation/PullbackCloner.cpp#L1951-L1961

Do you have any idea why `destroy_addr` isn't needed here? What happens if you add `diffBuilder. emitDestroyAddrAndFold(diffLoc, alloc);`?

efremale replied:
Not sure, we should probably add more tests with `Tracked<Float>` as in https://github.com/apple/swift/pull/33218

dan-zheng replied:
Adding tests with `Tracked<Float>` is a great idea!

dan-zheng commented on test/AutoDiff/SILOptimizer/forward_mode_diagnostics.swift:
```diff
@@ -85,10 +85,8 @@ func activeInoutParamControlFlow(_ array: [Float]) -> Float {
   return result
 }
 
-// FIXME(TF-984): Forward-mode crash due to unset tangent buffer.
```
There are more `// FIXME(TF-984)` references in the autodiff test suite, could you please handle all of them? You can find them via `grep -nr "TF-984" test/AutoDiff`.

rxwei reviewed (commented):
"Function-local allocations" were introduced in pullback generation so that adjoint buffers would dominate all subsequent use sites in the reverse CFG.  I don't see how they are needed in differential generation since the CFG is almost the same as the original function, so all tangent values we emit should have the same scope as their corresponding original value.

rxwei commented:
To me there seems to be a simpler, spot-on fix: For any non-wrt `inout` parameter that you treat as the result, create an adjoint buffer in the beginning of the function.  The "function-local allocation" fix applied in this patch seems like a big overkill because no other case would need function local allocations.

efremale committed:
[AutoDiff] Fix forward-mode crashes related to tangent buffers

Fixes TF-984, SR-13447.

Fixes foward-mode crashed related to:
- Missing tangent buffers for non-wrt `inout` parameters
- Tangent buffers not being initialized due to the corresponding
original buffer intialization instructions being non-active
- Non-varied indirect results not being initialized
- `emitDestroyValue` crashes due to `TangentVector` value category mismatch

rxwei commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -77,6 +77,9 @@ class JVPCloner::Implementation final
   /// tangent values.
   llvm::DenseMap<SILValue, AdjointValue> tangentValueMap;
 
+  enum IsInitialization_t : bool { NotInitialized, Initialized };
```
Did you mean `IsInitialized_t`?

That said, I feel the data structure could be simplified to a `llvm::SmallPtrSet<ValueBase *>`, and you can define a method `bool isTangentBufferInitialized(SILValue value)` to access this info.

efremale replied:
Yes, it should be `IsInitialized_t`.

I think having `map` instead of `set` is a bit safer in the sense that `map` allows us to explicitly track initialization (`true`/`false`) of every observed tangent buffer. This makes it possible to check whether a tangent buffer was previously observed (by `assert(map.count(tanBuf)`), which seems like a good check for correctness/debugging purposes.

dan-zheng replied:
Keeping `DenseMap` for explicit checking sounds pretty good to me.

rxwei replied:
Ok, sounds good.

rxwei commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -894,11 +947,17 @@ class JVPCloner::Implementation final
         bai->getLoc(), tanSrc, bai->getAccessKind(), bai->getEnforcement(),
         bai->hasNoNestedConflict(), bai->isFromBuiltin());
     setTangentBuffer(bb, bai, tanDest);
+    tangentBufferInitializationInfo[tanDest] = Initialized;
```
`tanDest` isn't initialized at this point, since `begin_access` doesn't change the initializedness of its operand. Plus, there can be multiple users of the result of `begin_access`, so setting this to `Initialized` before it is actually initialized feels unsafe.

rxwei commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -894,11 +947,17 @@ class JVPCloner::Implementation final
 
-    auto tanSrc = getTangentBuffer(bb, bai->getSource());
+    auto isInit = bai->getAccessKind() == SILAccessKind::Init;
+    auto tanSrc = getTangentBuffer(bb, bai->getSource(),
```
Since `begin_access` doesn't have any actual effect on the buffer, I feel we should pass `InitializationNotRequired` regardless here, and make the `tanDest` inherit the initializedness of `tanSrc` in this visitor. Then, we pass `InitializationRequired` in the visitor that *actually initializes* the buffer. Finally when we visit `end_access`, we assign the initializedness of the operand back to the source buffer.

rxwei commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -826,6 +869,7 @@ class JVPCloner::Implementation final
     auto &tanValDest = getTangentBuffer(si->getParent(), si->getDest());
     diffBuilder.emitStoreValueOperation(loc, tanValSrc, tanValDest,
                                         si->getOwnershipQualifier());
+    tangentBufferInitializationInfo[tanValDest] = Initialized;
```
When the store ownership qualifier is `Init`, I think you'll need to mark `tanValSrc` as uninitialized.

rxwei replied:
Also, according to [`store`](https://github.com/apple/swift/blob/master/docs/SIL.rst#id88)'s reference:
> `store %0 to %1 : $*T`
> If `%1` already references a value that requires release or other cleanup, that value must be loaded before being stored over and cleaned up.

We may need to switch on the initializedness of `tanValDest`. if it's already initialized, we call `destroy_addr` on it before emitting `store`.

efremale replied:
Is that true that `Init` uninitializes src? That looks unexpected.

rxwei replied:
Actually nvm, I mistook it for `copy_addr`. Since the source is an object (not an address), it needs no further handling.

rxwei commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -920,7 +980,10 @@ class JVPCloner::Implementation final
-    bufferMap.try_emplace({asi->getParent(), asi}, mappedAllocStackInst);
+    auto insertion =
+        bufferMap.try_emplace({asi->getParent(), asi}, mappedAllocStackInst);
+    if (insertion.second)
```
Should we instead `assert(insertion.second)`?

efremale replied:
Replaced all the `bufferMap.try_emplace` calls with `setTangentBuffer` (which is exactly `bufferMap.try_emplace` + `assert(inserted.second)`. It seems `bufferMap.try_emplace` uses were a bug.

rxwei commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -1061,7 +1130,9 @@ class JVPCloner::Implementation final
           diffBuilder.createTupleExtract(loc, tanSource, tanIndex, tanType);
-      bufferMap.try_emplace({tei->getParent(), tei}, tanBuf);
+      auto insertion = bufferMap.try_emplace({tei->getParent(), tei}, tanBuf);
+      if (insertion.second)
```
assert?

rxwei commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -1006,11 +1073,13 @@ class JVPCloner::Implementation final
         diffBuilder.createStructElementAddr(loc, tanOperand, tanField);
     // Update tangent buffer map for `struct_element_addr`.
     setTangentBuffer(bb, seai, tangentInst);
+    tangentBufferInitializationInfo[tangentInst] = Initialized;
```
`struct_element_addr` is valid on an uninitialized buffer. So I would probably let `tangentBufferInitializationInfo ]tangentInst]` inherit from `tangentBufferInitializationInfo[tanOperand]`.

rxwei commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -1061,7 +1130,9 @@ class JVPCloner::Implementation final
     } else {
       tanBuf =
```
This code (not your changes) feels very wrong in the first place... How is it a buffer? `tuple_extract` would never produce a buffer...

rxwei replied:
I feel this could be my code.  This should be calling `setTangentValue` on the extracted value instead, since the result has an object type, not an address.  Maybe it's not tested.

rxwei replied:
Thanks for fixing this for me!  Could you add a test case?  Maybe something like the following will trigger this code:
```swift
differential(at: 0.0) { x in
    let tuple = (x, 0.0)
    return tuple.x
}
```

rxwei commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -1475,18 +1551,37 @@ void JVPCloner::Implementation::prepareForDifferentialGeneration() {
+      auto tangentSILType = getRemappedTangentType(origSILType);
+      auto *tanBuf = differentialBuilder.createAllocStack(
+          differential.getLocation(), tangentSILType);
+      setTangentBuffer(origEntry, origParam, tanBuf);
```
Random idea: since `setTangentBuffer` is always called along with setting `tangentBufferInitializationInfo`, how about adding a parameter to `setTangentBuffer`? That way we could avoid potential invariant errors caused by not setting `tangentBufferInitializationInfo`.

```swift
setTangentBuffer(origEntry, origParam, tanBuf, NotInitialized)
```

efremale replied:
Good idea, added that: now we can provide an (optional) `IsInitialized_t` to `setTangentBuffer` to set `tangentBufferInitializationInfo`.

rxwei replied:
When would the caller let it default to `None`?

efremale replied:
The caller would let it default to `None` in cases when the initializedness is not known yet/will be set later.

However, there is only one place in the code like that right now in `prepareForDifferentialGeneration`:
```
...
for (auto &origBB : *original)
    for (auto i : indices(origIndResults))
      setTangentBuffer(&origBB, origIndResults[i], diffIndResults[i]); // it is pointless to set initializedness here since we'll do it for all `diffIndResults` immediately afterwards
  for (auto &diffIndResult : diffIndResults)
    tangentBufferInitializationInfo[diffIndResult] = NotInitialized;
```

efremale replied:
BTW, do you understand the purpose of calling `setTangentBuffer(&origBB, origIndResults[i], diffIndResults[i])` for every BB in `original`? Won't it be sufficient to just set it for the entry block?
```
for (auto i : indices(origIndResults))
  setTangentBuffer(differential.getEntryBlock(), origIndResults[i], diffIndResults[i]);
```

dan-zheng replied:
> BTW, do you understand the purpose of calling `setTangentBuffer(&origBB, origIndResults[i], diffIndResults[i])` for every BB in `original`? Won't it be sufficient to just set it for the entry block?

`setTangentBuffer(&origBB, origIndResults[i], diffIndResults[i])` is called for every original basic block in anticipation of future control flow support. It would sufficient to set it just for the entry block for now, but I think it's better to keep the existing logic to be future-facing.

rxwei replied:
IMHO there isn't a great reason to allow initialization info to be `Optional` in `setTangentBuffer`, because the programming interface should in principle enforce the invariants which all call sites make assumptions about. All `VJPCloner` visitor methods plus other places currently assume that every tangent buffer has an associated initialization info. Allowing initialization info to be `Optional` feels like an ad-hoc solution which on one hand specifically caters to `prepareForDifferentialGeneration` while on the other hand making all other call sites error-prone. One can easily forget to pass initialization info to `setTangentBuffer` when adding a new visitor in the future.

Since `JVPCloner` is emitting all tangent buffers (including in `prepareForDifferentialGeneration`), maybe we can delay the call to `setTangentBuffer` to the place where we **actually know** whether it is initialized or not?

efremale committed:
Fix bugs, add more tests, set initialization info in `setTangentBuffer`

rxwei commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -779,7 +826,7 @@ class JVPCloner::Implementation final
     auto loc = dvi->getLoc();
     auto tanVal = materializeTangent(getTangentValue(dvi->getOperand()), loc);
-    diffBuilder.emitDestroyValue(loc, tanVal);
+    diffBuilder.emitDestroyOperation(loc, tanVal);
```
Should we use `emitDestroyValueOperation`? IIRC `emitDestroyOperation` is for both buffers and objects. Here we are only dealing with objects.

rxwei commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -826,6 +875,9 @@ class JVPCloner::Implementation final
     auto &tanValDest = getTangentBuffer(si->getParent(), si->getDest());
     diffBuilder.emitStoreValueOperation(loc, tanValSrc, tanValDest,
                                         si->getOwnershipQualifier());
+    if (si->getOwnershipQualifier() == StoreOwnershipQualifier::Init)
+      tangentBufferInitializationInfo[tanValSrc] = NotInitialized;
```
Since `tanValSrc` is not a buffer, are these two lines redundant? Should we define a method for setting buffer initialization info such that it asserts that the argument's value category is always an address? That could prevent erroneous additions to the dictionary.
```cpp
void setTangentBufferInitializationInfo(SILValue buffer)
```

efremale replied:
Yes, these two lines are wrong!

The disadvantage of adding `setTangentBufferInitializationInfo` is that it is less efficient since it will require one more lookup in the map where we could avoid it. For example there are some pieces of code like:
```
auto &initInfo = tangentBufferInitializationInfo[tanBuf]; // map lookup
if (initInfo == IsIntitialized) {
  ...
}
initInfo = NotInitialized; // no need for another lookup
```
vs.
```
auto initInfo = getTangentBufferInitializationInfo(tanBuf); // map lookup
if (initInfo == IsIntitialized) {
  ...
}
setTangentBufferInitializationInfo(tanBuf, NotInitialized); // map lookup for the same key
```

It's probably very insignificant, but `setTangentBufferInitializationInfo` doesn't seem to be very useful. What do you think?

rxwei replied:
I see where you are coming from, but the point of the suggested change is mainly about enforcing the invariant that all tangent buffers have an initialization info, which will help ensure the correctness of the algorithm and IMHO outweighs potential redundant lookups. Moreover, the redundant lookup can be avoided entirely if we make `getTangentBufferInitializationInfo` return a reference so that you can modify `initInfo` directly. Given that the redundant lookup can be avoided entirely, this enforcement to prevent algorithm errors seems strictly positive to me.

rxwei commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -846,11 +899,17 @@ class JVPCloner::Implementation final
+    auto isInitialization = cai->isInitializationOfDest();
+    auto tanDest = getTangentBuffer(bb, cai->getDest(),
+                                    isInitialization ? InitializationNotRequired
+                                                     : InitializationRequired);
```
Since the destination buffer is not being read by `copy_addr`, passing `InitializationRequired` here can cause the buffer to be redundantly initialized, regardless of `isInitialization`. How about optimizing this by:
- Always passing `InitializationNotRequired` to `getTangentBuffer`.
- Before we call `createCopyAddr`, we use the tangent buffer initialization info to determine whether we pass `IsInitialization` or `IsNotInitialization` to `createCopyAddr`.

efremale committed:
Minor fixes

efremale committed:
Add explanatory comments

efremale committed:
Minor fixes: add `getTangentBufferInitializationInfo` and `setTangentBufferInitializationInfo`, make `initInfo` non-optional in `setTangentBuffer`

rxwei reviewed (approved):
Thanks for fixing this and addressing the comments!

dan-zheng commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -315,19 +315,28 @@ class JVPCloner::Implementation final
   }
 
+  IsInitialized_t &getTangentBufferInitializationInfo(SILValue tangentBuffer) {
+    auto it = tangentBufferInitializationInfo.find(tangentBuffer);
```
How about also adding an assertion in `getTangentBufferInitializationInfo`?
```cpp
    assert(tangentBuffer->getType().isAddress());
```

efremale replied:
It should already fail on `assert(it != .end())` since `assert(tangentBuffer->getType().isAddress())` is done on insertion to the map.

dan-zheng reviewed (approved):
Great fixes!

dan-zheng commented:
@swift-ci Please test

efremale committed:
Replace initializedness-tracking approach with an approach based on visiting all (un)initializing instructions where at least one parameter is active

We found out that it is very untrivial to correctly check initializedness of code with aggregate projections
(e.g. nested structs with multiple `struct_element_addr` referring to the same element). Instead of checking
initializedness, we can rely on correctness of the original function SIL and reproduce (un)initializing
instructions based on the original SIL.

dan-zheng committed:
NFC: minor cleanup.

Stylistic edits.
Fix test by removing invalid `@_silgen_name` usage.

dan-zheng committed:
Move tests to correct location, expose failing tests.

End-to-end forward-mode tests should be put in test/AutoDiff/validation-test.
The forward-mode differentiation changes do not fix all crashers.

dan-zheng committed:
Progress towards fixing tests.

efremale committed:
Fix failing tests

dan-zheng committed:
Fix iteration over original results.

Iterate over original differentiability result indices instead of directly
iterating over original indirect results and non-wrt `inout` parameters, which
may include results that are not relevant for differentiation.

Other cleanup included.

dan-zheng commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -1475,18 +1532,41 @@ void JVPCloner::Implementation::prepareForDifferentialGeneration() {
+
+  // Tangent buffers for non-varied original indirect results have to be
+  // initialized.
+  for (auto i : indices(origIndResults))
+    if (!activityInfo.isVaried(origIndResults[i], getIndices().parameters))
+      emitZeroIndirect(diffIndResults[i]->getType().getASTType(),
+                       diffIndResults[i], differential.getLocation());
```
Rather than directly iterating over original indirect results, we should iterate over differentiability result indices. This avoids including original indirect results that are not relevant for differentiation.

dan-zheng replied:
Fixed in https://github.com/apple/swift/pull/33633/commits/d37e2f46c047777814e4a3f92b20907fd5795969.

dan-zheng commented on lib/SILOptimizer/Differentiation/JVPCloner.cpp:
```diff
@@ -1475,18 +1532,41 @@ void JVPCloner::Implementation::prepareForDifferentialGeneration() {
 #endif
-  assert(origIndResults.size() + numNonWrtInoutParameters == diffIndResults.size());
+  differentialBuilder.setInsertionPoint(differential.getEntryBlock());
+  unsigned j = 0;
```
`unsigned j` doesn't appear to be used.

The logic for iterating over non-wrt `inout` parameters can be merged with the logic for iterating over original indirect results below: both should be subsumed by iterating over differentiability result indices.

dan-zheng replied:
Fixed inhttps://github.com/apple/swift/pull/33633/commits/d37e2f46c047777814e4a3f92b20907fd5795969.

dan-zheng commented:
@swift-ci Please test
